API Lesson 1 Pt 3: Update lesson to reflect latest version of Hello-Books-API (#144)
Ada-Developers-Academy/core-unit-2: api-1-setup-read/flask-setup.md

Reviewer: For running the server in debug mode, didn't Kelsey recommend a new command for that rather than setting FLASK_ENV?

|--|--|
|`01a-intro-to-flask` |`01b-flask-setup`|

## Flask Projects are Python Projects

> As we start writing APIs, we should remember that Flask projects are themselves Python projects. Like any other Python project, we need to manage the packages our project depends on—its dependencies. Like other Python projects, we handle this by creating a `venv` in which to install the packages described in the `requirements.txt` file. Let's briefly review the steps we take to perform this setup.

Author: @anselrognlie I removed the "Manage Dependencies" section and then incorporated your feedback about discussing Flask projects. 

Does this flow look ok?

Reviewer: Looks good. I also like the renaming of the section. Nothing special here. Flask projects are just Python projects!

One thing is that now that I read it in context, the "briefly review" here sounds repetative with the "brief review" in the first paragraph of the next section. How about changing this one to "revisit"?

```suggestion
As we start writing APIs, we should remember that Flask projects are themselves Python projects. Like any other Python project, we need to manage the packages our project depends on—its dependencies. Like other Python projects, we handle this by creating a `venv` in which to install the packages described in the `requirements.txt` file. Let's revisit the steps we take to perform this setup.
```

## Manage Dependencies

> Whenever we work on a Python project, we need to consider how we manage our dependencies which are the third party packages used by our project.

Reviewer: But by the  time students see this section, they will be using python 3.12, which will prevent them from pip installing system-wide at all, and we'll have started students using venvs as "projects" in precourse (and they'll have seen it on all the unit 1 projects), so I think this segue text could do with a re-flow. The main things for it to touch on should be:
1. Flask projects are Python projects, just like the other projects we've done.
2. We use venvs to manage our Python projects.
3. We will use a venv to manage our Flask project.
4. Here's a reminder on how to set up a venv for a project when the requirements are supplied to us.

## Manage Dependencies

As we start writing APIs, we should remember that Flask projects are Python projects. Whenever we work on a Python project, we need to consider how we manage our dependencies, which are the third party packages used by our project.  

Just as we have been doing before, we will want to continue using virtual environments, or `venv`s, to manage our projects.

> Since `flask` is a package with many of its own dependencies, we will need to use a `venv` to manage our Flask project so we do not make a mess of our system-wide Python installation. 

Author: @anselrognlie is this what you had in mind wrt your comment:

"Flask projects are Python projects, just like the other projects we've done.
We use venvs to manage our Python projects.
We will use a venv to manage our Flask project.
Here's a reminder on how to set up a venv for a project when the requirements are supplied to us."

?

Reviewer: I meant for that flow of topics to replace the entire existing chunk of content from Manage Dependencies to Virtual Environment.

The "Since `flask` is a package" and "In reality, no Python project" paragraphs serve no useful purpose, because we're going to get in the habit of using venvs for everything right away, so dwelling on the fact that it has its own dependencies doesn't inform why we will make a venv.

I'd stay away from phrasing best practices as "we will want to". Stick with we will, or we should. Leave the wanting aspect out of it. It hink this whol section could be boiled down to a single paragraph. Something like:

>As we start writing APIs, we should remember that Flask projects are themselves Python projects. Like any other Python project, we need to manage the packages our project depends on—its dependencies. Like other Python projects, we handle this by creating a `venv` in which to install the packages described in the `requirements.txt` file. Let's briefly review the steps we take to perform this setup.

> Here is a sample `requirements.txt` for a Flask project:

Reviewer: In the previous two paragraphs, can we remove the "will"s? It's unnecessary and complicates the tense.

```
alembic==1.13.1
blinker==1.7.0
click==8.1.7
Flask==3.0.2
Flask-Migrate==4.0.5
Flask-SQLAlchemy==3.1.1
greenlet==3.0.3
iniconfig==2.0.0
itsdangerous==2.1.2
Jinja2==3.1.3
Mako==1.3.2
MarkupSafe==2.1.5
packaging==23.2
pluggy==1.4.0
psycopg2-binary==2.9.9
pytest==8.0.0
python-dotenv==1.0.1
SQLAlchemy==2.0.25
typing_extensions==4.9.0

> Werkzeug==3.0.1

Reviewer 2: That looks good, better to show the new versions over older/possibly deprecated versions.

>    ### Common Debugging: Need to `$ pip install --upgrade pip`

Reviewer: I'd suggest condensing the callout titles. The titls on the one for upgrading `setuptools` now wraps (maybe due to changes in the Learn stylesheet?), leaving the info icon orphaned and looking strange. But if we rewrite that one, we should rewrite the others for consistency.
```suggestion
   ### Common Debugging: Upgrade `pip`
```

(I can't comment on the others since they aren't close enough to a modified region)

> │   ├── models

Reviewer: Another errant "will" in the sentence above this treeview. provide → provides

│   │   └── __init__.py
│   ├── __init__.py
│   └── routes
│       └── __init__.py
├── README.md
└── requirements.txt
```

### Routes

> The `app/routes` directory will be responsible for holding our routes. We should organize our routes so that all the routes corresponding to a particular model will be written in a single file. For example, all the endpoints related to the Book model would be written in a file called `book_routes.py`.

Reviewer 2: Not strictly necessary, but I was thinking the class name should be formatted as well.

```suggestion
The `app/routes` directory will be responsible for holding our routes. We should organize our routes so that all the routes corresponding to a particular model will be written in a single file. For example, all the endpoints related to the `Book` model would be written in a file called `book_routes.py`.
```

> The `app/routes` directory will be responsible for holding our routes. We should organize our routes so that all the routes corresponding to a particular model will be written in a single file. For example, all the endpoints related to the Book model would be written in a file called `book_routes.py`.

Reviewer: Just pointing out that I'm not _always_ against more future/speculative tenses. While this _could_ be expressed in simple present (The `app/routes` directory is responsible), it also works with "will be" since we're discussing a hypothetical structure.

The start-up logic is responsible for locating and applying any app configuration, and getting the server ready to receive requests.

> To get started, we will include the following code in `__init__.py`. This is the starting, boilerplate code to start a Flask application. Follow your curiosity if you would like to learn more about [`create_app`](https://flask.palletsprojects.com/en/3.0.x/patterns/appfactories/)

Reviewer 2: It isn't letting me leave a comment on the line because it isn't changed, but we should update the `create_app` function below to remove the `test_config` parameter since we do not add that until pt. 6 with testing.

    return app
```

> Configurations for the app can include things like, "Where's the location of our database?," "How do we load different data models, the objects that represent our data?," or "How can we set up template views, called Blueprints?"

Reviewer 2: It might be helpful to do a quick run through of the parent ticket for Part 1 to make sure everything was hit, on line 266 below we need to change the instructions for running Flask in debug mode to `flask run --debug`

Author: Oh yeah, good point, thanks for the reminder. I'll have a look at the parent ticket in addition to yours and Ansel's review comments.

> Configurations for the app can include things like, "Where's the location of our database?," "How do we load different data models, the objects that represent our data?," or "How can we set up template views, called Blueprints?"

Reviewer: In the prior code, we chould leave off the `test_config=None`

Reviewer: I'm not a fan of the blueprints line
"How can we set up template views, called Blueprints?"

That sounds like it was written before we had a good idea of what all the pieces in Flask were for. It also (at least for a Flask app) doesn't feel very "configuration-y" to me. Really, the only one of the three that sounds like configuration is locating the database, but I guess we're just going for a high-level view of the chunks of code that generally end up in `create_app`.

Anyway, if we want to keep a third thing here, maybe use something like "How do we run different logic in response to different requests?"

> ### !callout-danger

Reviewer: This callout feels very out of place to me. In combination with the general weakness of the description of configuraiton we give above, I feel like this just muddies the waters and creates a potential worry where none is required. Can we cut this?

Reviewer: Were you going to cut the callout?

Author: ohhh i misread this and cut lines 234-235. Thanks for following up.

      Sometimes the server gets stuck. What do we do in that case?
   </summary>

If something goes wrong when shutting down the server, or if the Terminal where the server was running is somehow closed without stopping the server, the server may not release the address and port where it was running. If that happens, a new server will not be able to run using that address and port, and will fail to start. This may be reported with an error message resembling: 

```bash
Address already in use
Port 5000 is in use by another program. Either identify and stop that program, or start the server with a different port.
On macOS, try disabling the 'AirPlay Receiver' service from System Preferences -> Sharing.
```

> If we know that we have not started the Flask server then we should check that the AirPlay Receiver is disabled like the Terminal output suggests. If it is enabled, we should disable it so that our computer can use port 5000 for running the Flask server. 

Reviewer: Include a bit more detail about what to do.
```suggestion
If we know that we have not already started another Flask server then we should check that the AirPlay Receiver is disabled like the Terminal output suggests. If it is enabled, we should disable it by unchecking its 'On' checkbox, so that our computer can use port 5000 for running the Flask server. 
```

> ![Screenshot of System Preferences showing the AirPlay Receiver is disabled](../assets/building-an-api/system-preferences-airplay-receiver.png)
> 

Reviewer: Remove the following blank line and add two spaces after the image to improve the caption formatting.
```suggestion
![Screenshot of System Preferences showing the AirPlay Receiver is disabled](../assets/building-an-api/system-preferences-airplay-receiver.png)  
```

> _Fig. Screenshot of System Preferences showing the AirPlay Receiver is disabled._

Reviewer: Give a bit more direction in the caption
```suggestion
_Fig. Screenshot of System Preferences with the `On` checkbox for AirPlay Receiver unchecked, indicating the AirPlay Receiver is disabled._
```

> Python  31756  user    3u  IPv4 0x92416cca3d5db1fb      0t0  TCP localhost:commplex-main (LISTEN)

Reviewer: We may want to update the error message that's shown. Newer Flask versions also detect whether they are on macos and warn about that Apple service that grabs port 5000. We may want to include information here about disabling that service if we haven't included it elsewhere.

Reviewer: Thanks for adding this!

```

> We can use the number labelled as `PID` (Process Identifier) to terminate the other running server with the command `kill -9 31756`, where we would replace `31756` with the `PID` value we got from running `lsof`.

Reviewer 2: Love adding the full term here for clarity ❤️

</details>

> 

Reviewer: In the server logs section below, I've had many students get confused about what's going on in the screen shot (they try to type in some of what's actually a part of the log output, and don't see anything happen.

Could we add a Figure caption that this is Flask log out in response to two web requests, and expand the text explanation so the steps to produce something similar are more clear?

These are the logs immediately after a `GET` request to `localhost:5000/i-didnt-define-this-endpoint-in-my-server-code`, which produced a `404` response.

> ![Screenshot of Terminal, which includes Flask server startup messages, and output for request/response of a 404 error](../assets/building-an-api/flask-setup_server-logs-404.png)
> 

Reviewer: Remove the following blank line and add two spaces after the image to force Learn to treat the caption as inline with the impage, but wrapped to the next line.
```suggestion
![Screenshot of Terminal, which includes Flask server startup messages, and output for request/response of a 404 error](../assets/building-an-api/flask-setup_server-logs-404.png)  
```
